django/utils/cache.py

```diff
@@ -131,72 +131,102 @@ def _not_modified(request, response=None):
-        # consider the list of values.
-        try:
-            etags = parse_etags(if_none_match or if_match)
-        except ValueError:
-            # In case of an invalid ETag, ignore all ETag headers.
-            # Apparently Opera sends invalidly quoted headers at times
-            # (we should be returning a 400 response, but that's a
-            # little extreme) -- this is bug #10681.
-            if_none_match = None
-            if_match = None
-
-    # If-None-Match must be ignored if original result would be anything
-    # other than a 2XX or 304 status. 304 status would result in no change.
-    # http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.26
-    if response and not (200 <= response.status_code < 300):
-        if_none_match = None
-        if_match = None
-
-    # If-Modified-Since must be ignored if the original result was not a 200.
-    # http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.25
-    if response and response.status_code != 200:
-        if_modified_since = None
-        if_unmodified_since = None
-
-    if not ((if_match and if_modified_since) or
-            (if_none_match and if_unmodified_since) or
-            (if_modified_since and if_unmodified_since) or
-            (if_match and if_none_match)):
-        # We only get here if no undefined combinations of headers are
-        # specified.
-        if ((if_none_match and (etag in etags or '*' in etags and etag)) and
-                (not if_modified_since or
-                    (last_modified and if_modified_since and last_modified <= if_modified_since))):
-            if request.method in ('GET', 'HEAD'):
-                return _not_modified(request, response)
-            else:
-                return _precondition_failed(request)
-        elif (if_match and (
-            (not etag and '*' in etags) or (etag and etag not in etags) or
-            (last_modified and if_unmodified_since and last_modified > if_unmodified_since)
-        )):
-            return _precondition_failed(request)
-        elif (not if_none_match and request.method in ('GET', 'HEAD') and
-                last_modified and if_modified_since and
-                last_modified <= if_modified_since):
+    if_none_match_etags = parse_etags(request.META.get('HTTP_IF_NONE_MATCH', ''))
+    if_modified_since = request.META.get('HTTP_IF_MODIFIED_SINCE')
+    if if_modified_since:
+        if_modified_since = parse_http_date_safe(if_modified_since)
+
+    # Step 1 of section 6 of RFC 7232.
+    if (if_match_etags and not _if_match_passes(etag, if_match_etags)):
+        return _precondition_failed(request)
+
+    # Step 2.
+    if (not if_match_etags and if_unmodified_since and
+            not _if_unmodified_since_passes(last_modified, if_unmodified_since)):
+        return _precondition_failed(request)
+
+    # Step 3.
+    if (if_none_match_etags and not _if_none_match_passes(etag, if_none_match_etags)):
+        if request.method in ('GET', 'HEAD'):
```

Comment:
Looks like a few tests are missing. If I remove this if/else and just leave `return _not_modified(request, response)`, no tests fail.